ENH: UDF operations for dask backend

gerrymanoim opened the pull request:
This PR adds support for udf operations to the dask backend. It unskips all non-window tests in

- `ibis/backends/tests/test_vectorized_udf.py`
- `ibis/backends/dask/tests/test_udf.py` (which were ported from the pandas backend. I've purposely left in xfail window tests so I can add to the tracker. )

In addition, I've built out the dask backend clones of `coerce_to_output` and `coerce_to_dataframe` which get used in udf/section operations. 

Post Merge:

- [ ] Update https://github.com/ibis-project/ibis/issues/2553 post merge (for things that are no longer skipped + more things skipped due to windowing).

jreback commented on ibis/backends/dask/__init__.py:
```diff
@@ -12,8 +12,9 @@
+from .udf import udf
 
-__all__ = ('connect', 'dialect', 'execute')
+__all__ = ('connect', 'dialect', 'execute', 'udf')
```
is this similar for spark/pandas (e.g. where its exposed)?

gerrymanoim replied:
Matches what the pandas backend exposes: https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/__init__.py#L13

icexelloss replied:
Can you use the API under https://github.com/ibis-project/ibis/blob/master/ibis/udf/vectorized.py#L102?

gerrymanoim replied:
This just creates an alias to that API: https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/udf.py#L137. 

Happy to do anything - was just working off the pandas API as a template.

icexelloss replied:
I am leaning towards using the API under udf.vectorized, it's more standard  since these UDFs works cross backends.

gerrymanoim replied:
I actually feel that we should keep the pandas and dask interfaces the same.

icexelloss replied:
IMHO having the code between Spark, Pandas, Dask the same is better than keeping Pandas and Dask the same but different from Spark.

The reason that pandas still have backends.pandas.udf is purely for backward compatible reasons. We should really deprecate it in favor of the vectorized.udf interface.

jreback replied:
ok i agree at the end of the day we want the same interface for all backends. so if `udf.vectorized` is is, then let's do that (@gerrymanoim if you wouldn't mind creating an issue for deprecating the old pandas udf interface)

jreback commented on ibis/backends/dask/udf.py:
```diff
@@ -0,0 +1,65 @@
+        func = op.func
+
+        # all grouping keys must be identical
+        assert len(set(tuple(arg.index) for arg in args)) == 1
```
test for this?

gerrymanoim replied:
This got replaced with `assert_identical_grouping_keys` which is tested directly.

icexelloss commented on ibis/backends/dask/tests/test_udf.py:
```diff
@@ -0,0 +1,257 @@
+# TODO - move these to the top level conftest?
+# Why does pandas keep them here
+@pytest.fixture
+def df(npartitions):
```
There is shared test under https://github.com/ibis-project/ibis/blob/master/ibis/backends/tests/test_vectorized_udf.py

icexelloss commented on ibis/backends/dask/udf.py:
```diff
@@ -0,0 +1,86 @@
+    # 1) an ungrouped window,
+    # 2) an ungrouped Aggregate node, or
+    # 3) an ungrouped custom aggregation context
+    @execute_node.register(type(op), *(itertools.repeat(dd.Series, nargs)))
```
Why do we need both pandas.Series and dd.Series here?

gerrymanoim replied:
This was more out of an abundance of caution - now that we no longer share dispatchers with the pandas backend I wanted to make sure the series registration was also being made. I don't believe there's any real scenario where we're mixing dask Series and pandas Series objects through - so I could remove.

icexelloss replied:
yeah prefer to remove for clarify.

icexelloss commented on ibis/backends/dask/udf.py:
```diff
@@ -0,0 +1,86 @@
+        # we're performing a scalar operation on grouped column, so
+        # perform the operation directly on the underlying Series
+        # and regroup after it's finished
+        args_objs = [make_selected_obj(arg) for arg in args]
```
What's the type of args_objs here?

gerrymanoim replied:
`args_objs: List[Union[dd.DataFrame, dd.Series]]`. 

I'll add a return type annotation to https://github.com/ibis-project/ibis/blob/master/ibis/backends/dask/execution/util.py#L34.

icexelloss commented on ibis/backends/dask/udf.py:
```diff
@@ -0,0 +1,86 @@
+        # and regroup after it's finished
+        args_objs = [make_selected_obj(arg) for arg in args]
+        groupings = args[0].index
+        return func(*args_objs).groupby(groupings)
```
Does the func here takes dd.Series or pd.Series?

gerrymanoim replied:
dd.Series.

icexelloss replied:
This might be different from how vectorized UDF API works:

The vectorized UDF always passes a pandas data structure to the user func regardless of the backend. Therefore I believe the func here should take pd.Series instead of dd.Series.

gerrymanoim replied:
But then we'd have to force computation each time - defeating the point of using dask, no?

icexelloss replied:
Yeah `map_partitions` is similar to what I have in mind.

icexelloss replied:
For the spark backend, we use the API here to implement this:

https://docs.databricks.com/spark/latest/spark-sql/udf-python-pandas.html#series-to-scalar-udf

Not sure what's the best way to do in Dask, but I think either `groupby().agg()` with custom aggregation or `map_partitions` would make sense,

icexelloss commented on ibis/backends/dask/execution/util.py:
```diff
@@ -31,7 +32,7 @@ def register_types_to_dispatcher(
 
 
-def make_selected_obj(gs: SeriesGroupBy):
+def make_selected_obj(gs: SeriesGroupBy) -> Union[dd.DataFrame, dd.Series]:
```
Since we have dd.DataFrame, maybe just use dd.SeriesGroupBy to be consistent

gerrymanoim replied:
`dd.SeriesGroupBy` doesn't exist. `DataFrameGroupBy` and `SeriesGroupBy` are in `dask.dataframe.groupby`

icexelloss replied:
Ok

icexelloss commented on ibis/backends/dask/execution/util.py:
```diff
@@ -47,25 +48,103 @@ def make_selected_obj(gs: SeriesGroupBy):
+        if isinstance(result, dd.Series):
+            return result.rename(result_name)
+        elif isinstance(result, dd.core.Scalar):
+            # TODO - computation
```
What does this TODO mean?

gerrymanoim replied:
This just tracks where we're forcing the dask backend to perform computation (as it isn't always clear). Will be useful when we go back and look at performance.

icexelloss replied:
Can you please create an internal Jira to track this?

icexelloss commented on ibis/backends/dask/execution/util.py:
```diff
@@ -47,25 +48,103 @@ def make_selected_obj(gs: SeriesGroupBy):
+        elif isinstance(result, dd.core.Scalar):
+            # TODO - computation
+            return dd.from_pandas(
+                pd.Series(result.compute(), name=result_name), npartitions=1
```
I am not sure about this. In my mind this function should be a lazy function and not executing the data. Why do we need to do this here?

gerrymanoim replied:
This code was transferred to this function and already exists in the backend. 

The issue is that we can't easy wrap a `dd.core.Scalar` in another dask object without computing it. I have a TODO for myself to figure out if there's somewhere in upstream dask that we could do this.

icexelloss replied:
Ok. Let's track as follow up.

icexelloss commented on ibis/backends/dask/execution/util.py:
```diff
@@ -47,25 +48,103 @@ def make_selected_obj(gs: SeriesGroupBy):
+                pd.Series(result.compute(), name=result_name), npartitions=1
+            )
+        else:
+            series = dd.from_array(np.repeat(result, len(index)))
```
This could below up memory if index is too large?

jreback replied:
these are all Reduction ops, this is prob ok.

icexelloss replied:
Oh yeah that's true. Reductions should be fine.

icexelloss commented on ibis/backends/dask/execution/util.py:
```diff
@@ -47,25 +48,103 @@ def make_selected_obj(gs: SeriesGroupBy):
         return result.rename(result_name)
 
 
+def _coerce_to_dataframe(data: Any, names: List[str]) -> pd.DataFrame:
```
Return type is wrong?

icexelloss replied:
Can you add in the docstring what are the valid input types?

jreback replied:
definitly add Parameters / Returns as much as possible. wouldn't object to Examples if useful.

icexelloss commented on ibis/backends/dask/execution/util.py:
```diff
@@ -86,6 +165,13 @@ def safe_concat(dfs: List[Union[dd.Series, dd.DataFrame]]) -> dd.DataFrame:
 
     TODO - performance.
```
Can you expand this TODO? Not clear what performance are you referring to here

gerrymanoim replied:
Just general performance - this method works in that it produces the correct result, but if we start encountering performance issues this is definitely an area I would look at optimizing.

icexelloss replied:
kk, it'd be great if you can document what performance issue this function has. Just looking at this comment I have little idea why this function has performance issues.

gerrymanoim replied:
Added more specificity to the note.

icexelloss commented on ibis/backends/dask/execution/util.py:
```diff
@@ -142,3 +228,16 @@ def compute_sorted_frame(
+
+
+def assert_identical_grouping_keys(*args):
+    indicies = [arg.index for arg in args]
```
indices?

gerrymanoim replied:
ack - spelling - apologies.

icexelloss commented on ibis/backends/dask/execution/util.py:
```diff
@@ -142,3 +228,16 @@ def compute_sorted_frame(
+    # Depending on whether groupby was called like groupby("col") or
+    # groupby(["cold"]) index will be a string or a list
+    if isinstance(indicies[0], list):
+        indicies = [index[0] for index in indicies]
```
Can index contain multiple elements? groupby(["cold", "warm"])?

gerrymanoim replied:
Errr - yes actually. Sorry! Not sure why I thought they didn't? Will fix the logic here.

icexelloss replied:
Sounds good. Can you add a test for this?

icexelloss commented on ibis/backends/tests/test_vectorized_udf.py:
```diff
@@ -352,7 +345,7 @@ def test_elementwise_udf_destruct_exact_once(backend, alltypes):
         )
         def add_one_struct_exact_once(v):
-            key = v.iloc[0]
+            key = v.loc[0]
```
Why this change?

gerrymanoim replied:
Dask has no support for `iloc`.

icexelloss replied:
Hmm. See my other comments. The object passed to these UDFs should be pandas object not dask objects..

gerrymanoim replied:
Hmm. That's a good point - I need to think about this. Right now we're applying the function and trusting dask to handle it. 

I'm not sure how you apply a function across independent Series? There's no guarantee that they're partitioned the same or on the same workers.

icexelloss commented on ibis/backends/dask/execution/util.py:
```diff
@@ -47,25 +48,103 @@ def make_selected_obj(gs: SeriesGroupBy):
+    elif isinstance(data, dd.Series):
+        # This takes a series where the values are iterables and converts each
+        # value into its own row in a new dataframe.
+        first_val = data.head(1).values[0]
```
why do we need to read the first value?

icexelloss replied:
Looks like you only need the type?

gerrymanoim replied:
I need the length for `num_cols`

icexelloss commented on ibis/backends/dask/execution/util.py:
```diff
@@ -47,25 +48,103 @@ def make_selected_obj(gs: SeriesGroupBy):
+        # This takes a series where the values are iterables and converts each
+        # value into its own row in a new dataframe.
+        first_val = data.head(1).values[0]
+        num_cols = len(first_val)
```
Why would you have multiple cols for a dd.Series?

gerrymanoim replied:
This is not the length of the series, but the lenght of the iterable in `first_value`. We take each value in each iterable and create a new column. See the equivalent pandas code in: https://github.com/ibis-project/ibis/blob/0f3ef9af4979e60ccc8ab523d0b7ff1c0106051f/ibis/util.py#L100

Let me know if the variable name or comment itself is not clear - I'll update.

icexelloss replied:
I see. Yeah the code you referring to has this in docstring that this is a series of list/tuple. Can you make that clear here as well?

icexelloss replied:
Also, can we infer this from output schema of the ibis node instead of actual data?

gerrymanoim replied:
Great point! 

Fixed.

icexelloss commented on ibis/backends/dask/udf.py:
```diff
@@ -0,0 +1,121 @@
+        # and regroup after it's finished
+        args_objs = [make_selected_obj(arg) for arg in args]
+        groupings = args[0].index
+        return dd.map_partitions(func, *args_objs).groupby(groupings)
```
Why do we need a groupby for element wise UDF?

gerrymanoim replied:
https://github.com/ibis-project/ibis/pull/2648#discussion_r592864535

icexelloss replied:
Sorry I am not familiar with the part of the dask API used here:

- What are the types of `args_objs` here? 
- How does the freeform function dd.map_partitions() work
- Are we returning a ddgb.DataFrameGroupby object from this function? If so, why are we not returning a dd.DataFrame?

gerrymanoim replied:
- `make_selected_obj` in this case returns a `dd.Series`, so `args_objs` is a `Lis[dd.Series]`. 
- My understanding is `map_partitions`  sets up all the machinery to apply the func on the args passed. It knows about which args are partitionable vs which are scalars and pushes that down accordingly. Is that what you're asking?
- This matches what pandas does in https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/udf.py#L188

icexelloss replied:
Oh I see. Actually the pandas logic looks a bit puzzling as well.

For my knowledge, when this returns a SeriesGroupBy object, how does the down stream node consume it? i.e., is there a step somewhere else to turn this back to a dd.DataFrame if the down stream operation is sth like mutate?

gerrymanoim replied:
It looks like generally an aggregation is performed on it afterwards thus making it a ungrouped series post the aggregation. 
For example: `t.groupby(t.key).aggregate(my_add=my_add(t.b, t.c).sum())`

icexelloss commented on ibis/backends/dask/udf.py:
```diff
@@ -0,0 +1,121 @@
+
+    @execute_node.register(
+        ops.ElementWiseVectorizedUDF,
+        *(itertools.repeat(ddgb.SeriesGroupBy, nargs)),
```
Why would element wise UDF takes a SeriesGroupBy?

gerrymanoim replied:
Unless I'm missing something this matches what is registered for pandas: https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/udf.py#L171

icexelloss replied:
KK I think I remember now. It's valid to run element wise UDF inside groupby.

icexelloss commented on ibis/backends/dask/udf.py:
```diff
@@ -0,0 +1,121 @@
+    # Define an execution rule for a simple elementwise Series
+    # function
+    @execute_node.register(
+        ops.ElementWiseVectorizedUDF, *(itertools.repeat(pandas.Series, nargs))
```
Can we remove the signature with pd.Series? I don't think those are used here.

icexelloss commented on ibis/backends/dask/udf.py:
```diff
@@ -0,0 +1,121 @@
+        # kwargs here. This is true for all udf execution in this
+        # file.
+        # See ibis.udf.vectorized.UserDefinedFunction
+        return op.func(*args)
```
I think func takes pd.Series not dd.Series here?

icexelloss replied:
Let's add some assertion in the test case to verify input object types

gerrymanoim replied:
https://github.com/ibis-project/ibis/pull/2648#discussion_r592887643

icexelloss commented on ibis/backends/dask/udf.py:
```diff
@@ -0,0 +1,121 @@
+    @execute_node.register(type(op), *(itertools.repeat(pandas.Series, nargs)))
+    def execute_udaf_node_no_groupby(op, *args, aggcontext, **kwargs):
+        # not grouping - no aggctx
+        return op.func(*args)
```
Similar to above. func should take pd.Series not dd.Series.

gerrymanoim replied:
https://github.com/ibis-project/ibis/pull/2648#discussion_r592887643

gerrymanoim replied:
I understand where we got to with `elementwise` but this is another place where I'm not sure how this would work for partitioned datasets. For a function like:

```python
@analytic(input_type=[dt.double], output_type=dt.double)
def calc_zscore(s):
    return (s - s.mean()) / s.std()
```

If that needs to operate on actual pandas objects than I need to materialize the entire series, instead of having dask deal with it. I'm not sure how this would work if that's the actual requirement.

icexelloss replied:
analytics works on groups, so I'd imagine we'd use Dask to shuffle/repartition by groups an apply this function on each partition.

If there is no groupby key, i.e., call this UDF over the entire series then there is no other option than materialize the entire series. It is not a common use case and strongly discouraged for large dataset.

jreback commented on ibis/backends/dask/execution/util.py:
```diff
@@ -47,25 +48,103 @@ def make_selected_obj(gs: SeriesGroupBy):
+        first_val = data.head(1).values[0]
+        num_cols = len(first_val)
+        # NOTE - We add a detailed meta here so we do not drop the key index
+        # downstream. This seems to be fixed in versions of dask > 2020.12.0
```
we can always push the min version in ibis (separate PR)

gerrymanoim replied:
We probably want to reasonably soon - but I didn't want to do it at the same time.

jreback reviewed (commented):
didnt' get too far thru

icexelloss commented on ibis/backends/tests/test_vectorized_udf.py:
```diff
@@ -89,15 +89,15 @@ def overwrite_struct_reduction(v, w):
 
 
-@pytest.mark.only_on_backends(['pandas', 'pyspark'])
+@pytest.mark.only_on_backends(['pandas', 'pyspark', 'dask'])
```
Can you please add type assertions in the test UDFs here? e.g.

```
@analytic(
    input_type=[dt.double, dt.double],
    output_type=dt.Struct(['demean', 'demean_weight'], [dt.double, dt.double]),
)
def demean_struct(v, w):
    assert isinstance(v, pd.Series)
    assert isinstance(w, pd.Series)
    return v - v.mean(), w - w.mean()
```

gerrymanoim commented:
I think I addressed all the feedback. For the case of an ungrouped `ops.ReductionVectorizedUDF` I'm currently forcing computation because the downstream destructuring code can't work with a `dd.core.Scalar`. Given that we know the output will eventually be a `dt.Struct` I _think_ it should be possible to use `dask.bag` or something to represent a collection of individual dask scalars and get around this but I'm not sure.  Given that this isn't a common use case and strongly discouraged for large datasets (and this PR is already complicated)  I'd like to leave that to a follow on.

jreback commented on ibis/backends/dask/execution/aggregations.py:
```diff
@@ -77,7 +77,7 @@ def execute_aggregation_dataframe(
     for metric in op.metrics:
         piece = execute(metric, scope=scope, timecontext=timecontext, **kwargs)
-        piece = maybe_wrap_scalar(piece, metric)
+        piece = coerce_to_output(piece, metric)
```
is this a consistent name we use in other backends?

gerrymanoim replied:
Yep - I'm emulating the pandas code https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/execution/generic.py#L437

jreback commented on ibis/backends/dask/execution/selection.py:
```diff
@@ -117,13 +116,10 @@ def compute_projection_column_expr(
     )
 
     result = execute(expr, scope=scope, timecontext=timecontext, **kwargs)
+    result = util.coerce_to_output(result, expr, data.index)
```
nbd, but should consistently import `coerce_to_output` (or just use `util`, prefer the first)

jreback commented on ibis/backends/dask/execution/util.py:
```diff
@@ -47,25 +50,126 @@ def make_selected_obj(gs: SeriesGroupBy):
+        Optional. If passed, scalar results will be broadcasted according
+        to the index.
+
+    Returns
```
can you add some examples (for future you :->)

jreback commented on ibis/backends/dask/execution/util.py:
```diff
@@ -47,25 +50,126 @@ def make_selected_obj(gs: SeriesGroupBy):
+
+    Returns
+    -------
+    result: A Series or DataFrame
```
dd.Series & dd.DataFrame

gerrymanoim replied:
Hmmm....I can update the comment if you'd like, but I think the implicit assumption is inside the dask backend a `Series` or `DataFrame` is always referring to the dask versions.

jreback commented on ibis/backends/dask/execution/util.py:
```diff
@@ -47,25 +50,126 @@ def make_selected_obj(gs: SeriesGroupBy):
+    (1) A list/tuple of Series -> each series is a column
+    (2) A list/tuple of scalars -> each scalar is a column
+    (3) A Dask Series of list/tuple -> each element inside becomes a column
+    (4) dd.DataFrame -> the data is unchanged
```
examples!

jreback commented on ibis/backends/dask/execution/util.py:
```diff
@@ -47,25 +50,126 @@ def make_selected_obj(gs: SeriesGroupBy):
         return result.rename(result_name)
 
 
+def _coerce_to_dataframe(
```
explicitly private?

gerrymanoim replied:
Yeah...currently it is really only used as a helper to `coerce_to_output` for `ir.Destruct/ir.Struct` expressions.

jreback commented on ibis/backends/dask/execution/util.py:
```diff
@@ -47,25 +50,126 @@ def make_selected_obj(gs: SeriesGroupBy):
+        # downstream. This seems to be fixed in versions of dask > 2020.12.0
+        dtypes = map(ibis_dtype_to_pandas, types)
+
+        def make_meta_series(dtype):
```
would not object to making these functions module level private

jreback commented on ibis/backends/dask/execution/util.py:
```diff
@@ -47,25 +50,126 @@ def make_selected_obj(gs: SeriesGroupBy):
+            for i, dtype in enumerate(dtypes)
+        ]
+        result = dd.concat(series, axis=1)
+    elif isinstance(data, (tuple, list)):
```
blank line for readability

jreback commented on ibis/backends/dask/execution/util.py:
```diff
@@ -47,25 +50,126 @@ def make_selected_obj(gs: SeriesGroupBy):
+        ]
+        result = dd.concat(series, axis=1)
+    elif isinstance(data, (tuple, list)):
+        if isinstance(data[0], dd.Series):
```
what if data is empty?

jreback commented on ibis/backends/dask/execution/util.py:
```diff
@@ -47,25 +50,126 @@ def make_selected_obj(gs: SeriesGroupBy):
+    elif isinstance(data, (tuple, list)):
+        if isinstance(data[0], dd.Series):
+            result = dd.concat(data, axis=1)
+        else:
```
can you make this elifs?

jreback commented on ibis/backends/dask/execution/util.py:
```diff
@@ -47,25 +50,126 @@ def make_selected_obj(gs: SeriesGroupBy):
+            #     result = dd.from_array(np.array(data))
+    else:
+        raise ValueError(f"Cannot coerce to DataFrame: {data}")
+    result.columns = names
```
blank line

jreback commented on ibis/backends/dask/execution/util.py:
```diff
@@ -142,3 +252,16 @@ def compute_sorted_frame(
     return result
+
+
+def assert_identical_grouping_keys(*args):
```
is this a testing function?

gerrymanoim replied:
No - as you mentioned later, we perform this check in a few places and it is easier to put this function in a central place with a nice error message.

jreback commented on ibis/backends/dask/tests/execution/test_util.py:
```diff
@@ -0,0 +1,44 @@
+
+
+def test_identical_grouping_keys_assertion(df):
+    gdf = df.groupby("dup_strings")
```
might be worth a more in depth comment

jreback commented on ibis/backends/dask/tests/test_udf.py:
```diff
@@ -0,0 +1,483 @@
+        npartitions=npartitions,
+    )
+
+
```
might be worth a 1-liner of each of these as to the purpose (or a small rename)

jreback commented on ibis/backends/dask/tests/test_udf.py:
```diff
@@ -0,0 +1,483 @@
+)
+def quantiles(series, *, quantiles):
+    return list(series.quantile(quantiles))
+
```
wouldn't be averse to a line break here

```
-----------
Testing
------------
```

or similar (and one like this for Fixtures), nbd, but improves readability at a glance

jreback commented on ibis/backends/dask/tests/test_udf.py:
```diff
@@ -0,0 +1,483 @@
+    assert nullable(t.a.type()) == ()
+
+
+def test_udaf_parameter_mismatch():
```
usually prefer to put error tests together for easy readability (e.g. not in the middle of other types of testing)

gerrymanoim replied:
Mostly I copied/translated the pandas directly in order. 

I can reorder these.

jreback commented on ibis/backends/dask/udf.py:
```diff
@@ -0,0 +1,161 @@
+        func = op.func
+
+        # all grouping keys must be identical
+        assert_identical_grouping_keys(*args)
```
i c, you are using this here, ok then (assume you are testing a good error message)

jreback commented on ibis/backends/dask/udf.py:
```diff
@@ -0,0 +1,161 @@
+        # file.
+        # See ibis.udf.vectorized.UserDefinedFunction
+        if isinstance(op._output_type, dt.Struct):
+            # dask does not know how to glue together Tuple[*pd.Series].
```
is this a bug on the dask side?

gerrymanoim replied:
I don't know if this is a bug vs. us just using this in an unexpected way. 

```python

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB set_trace (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[56] > /home/g/dev/gerrymanoim/ibis/ibis/backends/dask/udf.py(68)execute_udf_node()
-> df.index = args[0].index
   5 frames hidden (try 'help hidden_frames')
(Pdb++) df
Dask Series Structure:
npartitions=2
0       object
3650       ...
7299       ...
dtype: object
Dask Name: wrapper, 6 tasks
(Pdb++) df.compute()
0         (1.0, 2.0)
1       (11.1, 12.1)
2       (21.2, 22.2)
3       (31.3, 32.3)
4       (41.4, 42.4)
            ...     
3645    (51.5, 52.5)
3646    (61.6, 62.6)
3647    (71.7, 72.7)
3648    (81.8, 82.8)
3649    (91.9, 92.9)
Length: 7300, dtype: object
(Pdb++) dd.map_partitions(op.func, *args, meta="object")
Dask Series Structure:
npartitions=2
0       object
3650       ...
7299       ...
dtype: object
Dask Name: add_one_struct, 6 tasks
(Pdb++) dd.map_partitions(op.func, *args, meta="object").compute()
0    ([1.0, 11.1, 21.2, 31.3, 41.4, 51.5, 61.6, 71....
1    ([21.2, 31.3, 41.4, 51.5, 61.6, 71.7, 81.8, 91...
```

By wrapping we get a result of the right "shape"

jreback commented on ibis/backends/dask/udf.py:
```diff
@@ -0,0 +1,161 @@
+            return apply_func(*cols)
+
+        # NOTE - We add a detailed meta here so we do not drop the key index
+        # downstream. This seems to be fixed in versions of dask > 2020.12.0
```
another reason we should have a higher min :->

gerrymanoim replied:
Yep - I can bump it in the next PR.

jreback commented on ibis/backends/tests/test_vectorized_udf.py:
```diff
@@ -76,6 +88,8 @@ def demean_struct(v, w):
     output_type=dt.Struct(['mean', 'mean_weight'], [dt.double, dt.double]),
 )
 def mean_struct(v, w):
```
are these tested for both inputs?

gerrymanoim replied:
Not sure what you mean by that. Could you clarify?

jreback replied:
so some udf's only receive Series some can receive ndarray & Series. where is this determined and is this tested (that can get either input)

gerrymanoim replied:
From talking to @icexelloss -  it actually seems to be a property of the window implementation of the backend, not anything to do with the UDF or what the user requests.

icexelloss replied:
Yeah this is an existing issue. Window aggregation UDF takes np.ndarray and groupby aggregation takes pd.Series.

jreback reviewed (changes requested):
looks good a few comments

icexelloss commented on ibis/backends/dask/tests/test_udf.py:
```diff
@@ -0,0 +1,483 @@
+@pytest.mark.xfail(
+    raises=NotImplementedError, reason='TODO - windowing - #2553'
+)
+def test_udaf_window(t2, df2):
```
It did feel like some of the tests here are similar to the test in https://github.com/ibis-project/ibis/blob/master/ibis/backends/tests/test_window.py#L375 for example

I'd suggest sharing as much of these as possible to the backends/tests so other backends can benefit from the same test.

gerrymanoim replied:
All these tests were lifted directly from the pandas backend which makes tracking feature parity a lot easier. I'm happy to further refactor those out of both backends, but I'd like to do that separately.

icexelloss replied:
Yeah that's fine. Ideally the more shared tests the better.

icexelloss commented on ibis/backends/dask/tests/test_udf.py:
```diff
@@ -0,0 +1,483 @@
+    assert nullable(t.a.type()) == (type(None),)
+
+
+def test_nullable_non_nullable_field():
```
I think this test can probably be lifted to expr? Don't need you need any Dask specific things to test this.

gerrymanoim replied:
https://github.com/ibis-project/ibis/pull/2648#discussion_r600751733

icexelloss commented on ibis/backends/dask/tests/test_udf.py:
```diff
@@ -0,0 +1,483 @@
+    tm.assert_series_equal(result, expected)
+
+
+def test_udaf_groupby(df3):
```
Perhaps can be follow up but ideally these should be backend shared tests.

gerrymanoim replied:
https://github.com/ibis-project/ibis/pull/2648#discussion_r600751733

icexelloss commented on ibis/backends/dask/udf.py:
```diff
@@ -0,0 +1,161 @@
+        meta = pandas.Series([], dtype=op._output_type.to_dask())
+
+        @dask.delayed
+        def lazy_agg(*series):
```
Is series here pd.Series or dd.Series?

gerrymanoim replied:
`pd.Series` - the assertions are now inside of every test udf function.

icexelloss replied:
Would be helpful if you can comment/annotate the input types to this function.

icexelloss commented on ibis/backends/dask/udf.py:
```diff
@@ -0,0 +1,161 @@
+            # computation is forced so that we can work with the scalar
+            # TODO computation
+            out_dtype = op._output_type.to_dask()
+            if out_dtype == object:
```
Perhaps better to check if  op._output_type is struct.

dask object type maybe use to present other types, like, string?

gerrymanoim replied:
Yeah - great point, thanks!

icexelloss commented on ibis/backends/tests/test_vectorized_udf.py:
```diff
@@ -14,16 +16,19 @@
 
 @elementwise(input_type=[dt.double], output_type=dt.double)
 def add_one(s):
+    assert isinstance(s, (np.ndarray, pd.Series))
```
I think we only need to loose the assertion on reduction to be `isinstance(s, (np.ndarray, pd.Series))`

elementwise and analytic only takes pd.Series

gerrymanoim replied:
I can try that out.

icexelloss commented on ibis/backends/dask/udf.py:
```diff
@@ -0,0 +1,161 @@
+        def lazy_agg(*series):
+            return aggcontext.agg(series[0], op.func, *series[1:])
+
+        lazy_result = lazy_agg(*args)
```
Discussed offline. Let's put some comments here to explain what is happening:

dask.delayed will turn dd.Series into pd.Series (by partition) and pass down to actual lazy_agg function.

icexelloss replied:
Per discussion this doesn't do map partitions, it just collect the entire dd.Series to a single worker as pd.Series

jreback commented:
also pls rebase @gerrymanoim

jreback commented on ibis/backends/dask/tests/execution/test_util.py:
```diff
@@ -0,0 +1,53 @@
+        assert_identical_grouping_keys(a, b, c)
+
+    # simple case for single grouping given as a List[str]
+    gdf = df.groupby(["dup_strings"])
```
can you parametrize this test?

jreback reviewed (changes requested):
looks good a comment and question

icexelloss reviewed (approved):
LGTM

jreback commented:
thanks @gerrymanoim very nice!

can prob update whatsnew at some point in a future PR
